sklearn/preprocessing/imputation.py

```diff
@@ -366,13 +397,37 @@ def transform(self, X):
 
             mask = _get_mask(X, self.missing_values)
             n_missing = np.sum(mask, axis=self.axis)
-            values = np.repeat(valid_statistics, n_missing)
 
-            if self.axis == 0:
-                coordinates = np.where(mask.transpose())[::-1]
+            if self.strategy == 'knn':
+                if self.axis == 1:
+                    X = X.transpose()
+                    mask = mask.transpose()
+                    statistics = statistics.transpose()
+                missing_index = np.where(mask.any(1))[0]
+                for i, row in zip(missing_index, X[missing_index]):
+                    col_index = np.where(np.logical_not(np.isnan(row)))[0]
+                    impute_index = np.where(np.isnan(row))[0]
+                    neigh = NearestNeighbors(self.kneighbor)
+                    neigh = neigh.fit(statistics[:, col_index])
+                    _dist, ind = neigh.kneighbors(row[np.logical_not(np.isnan(row))],
+                                           self.kneighbor)
+                    #tree = KDTree(statistics[:, col_index])
+                    #dist, ind = tree.query(row[np.logical_not(np.isnan(row))],
+                    #                       k=self.kneighbor)
+                    nn_index = ind[0]
+                    X[i][impute_index] = statistics[nn_index][:, impute_index].mean(0)
```

Comment:
If using nearest neighbors I think you can just use `KNeighborsRegressor` to perform the (weighted) mean for you.